### DISCUSSION TITLE
张量程序抽象有助于单一算子的聚焦，那对于全局的模型算子图来说这样的优化是最优吗？

### DISCUSSION THREAD

**Discussion by liu-mengyang** [mlc-ai/mlc-zh#60] [Q&A] [answered]
通过张量程序抽象，再逐个对每个算子进行针对性地张量函数优化，这一做法让我觉得更像是对一个复杂全局优化问题的简化，将一个问题以一种贪心的思想进行优化。
但这样的做法真的是最优吗？如果聚焦于全局模型算子一起优化是不是会更优呢？
换一种问法，这样的贪心可以被证明最优吗？或者不贪心，全局去做优化是否无法实现即不可行呢？

**Comment by yzh119** [comment, marked as answer]
很好的问题，如果每个算子的优化都是独立的话，那么每个算子调到最优就是全局最优解。

图层面的优化更加复杂，一方面需要考虑算子融合和各种子图变换，另一方面我们可能会对每个算子的输入输出做格式转化（例如array packing，NCHW <-> NHWC），这两者都必须考虑全局的信息。这也是为什么需要Relax这个层级的抽象。

**Reply by liu-mengyang** [reply]
谢谢解答！由于我目前才看到第二节课，很多东西特别是图优化的内容可能还不太了解，不过我还是有所疑惑。按您说的，所以图优化是先于算子优化的，我认为这其实也是一个贪心，贪心的最优性证明需要符合最优子结构和贪心选择性，这能被证明吗？
我再梳理想了想，也就是说如果能证明算子优化的变换对图优化过程不会有影响其实就可以了，那这确实会不存在影响吗？也就是说图优化的指标中不存在算子执行的内容，比如考虑单一算子的计算复杂度等等进行变换决策。

**Reply by yzh119** [reply]
google有做过算子级和图层面联合优化的工作：https://mangpo.net/papers/xla-autotuning-pact2021.pdf

> 按您说的，所以图优化是先于算子优化的，我认为这其实也是一个贪心，贪心的最优性证明需要符合最优子结构和贪心选择性，这能被证明吗？

图优化不一定要先于算子优化，可能是一个迭代优化的过程 图优化->算子优化->图优化->算子优化 etc

> 也就是说如果能证明算子优化的变换对图优化过程不会有影响其实就可以了，那这确实会不存在影响吗？

如果对输入输出做格式转换那么就有影响（因为格式转换本身也是一个算子）

**Reply by liu-mengyang** [reply]
了解了了解了，感谢！
